connections/hole-punching.md

```diff
@@ -0,0 +1,288 @@
+being no need for hole punching. The two hosts should use either TCP or QUIC to
+communicate directly.
+
+### Private Non-Browser (A) to Public Non-Browser (B)
+
+Given that B is publicly reachable, A can establish a direct TCP or QUIC
+connection to B.
+
+### Private Browser (A) to Public Non-Browser (B)
+
+Given that B is publicly reachable A can establish a direct connection, though
+multiple restrictions are imposed by the browser platform type of A:
+
+- The only transport protocol allowing direct connections on browser platforms
+  is WebSockets. Thus B has to support the WebSocket transport protocol.
+
+- Browser platforms do not allow insecure WebSocket connections, thus B has to
+  have a valid TLS certificate to offer secure Websocket communication.
+
+In cases where B does not fulfill both of these requirements, additional steps
+as detailed in [Private Browser (A) to Private Non-Browser
+(B)](#private-browser-a-to-private-non-browser-b) are necessary.
+
+### Public or Private Non-Browser (A) to Private Non-Browser (B)
+
+In order to establish a direct connection from A to B, one needs to utilize some
+means of firewall / NAT hole punching. Given the permissiveness of non-browser
+platforms there are many ways to achieve this goal, with the below being one of
+them.
+
+- B uses the observed addresses from the [Identify] protocol in combination with
+  the [AutoNAT] protocol to determine whether it is running behind firewalls
+  and/or NATs and in the latter case what type of NATs.
+  
+  Note: Hole punching fails if either A or B runs behind a [symmetric
+  NAT][symmetric-nat].
+
+- B establishes a TCP or QUIC connections to one or more relay servers and
+  listens for incoming connection requests via the [circuit relay v2
+  protocol][circuit-relay-v2]. Once B established a reservation at one or more
+  relay servers, B can advertise its relayed addresses (e.g.
+  `/ip4/.../tcp/.../p2p/QmRelay`) via some external mechanism.
+
+  Note on decentralization: Given the low resource requirements of the [circuit
+  relay v2 protocol][circuit-relay-v2] any public host can serve as a relay
+  server.
+
+- A discovers the relayed address of B through some external mechanism. Given
+  the relayed address, A can establish a relayed connection to B via the [circuit
+  relay v2 protocol][circuit-relay-v2] protocol over TCP or QUIC. A and B can
+  then run the [Direct Connection Upgrade through Relay][DCUTR] protocol over
+  the relayed connection to coordinate TCP or QUIC hole punching in the best
+  case establishing a direction connection.
+
+For more details see the [Project Flare][project-flare] project proposal.
+
+### Private Browser (A) to Private Non-Browser (B)
+
+As mentioned above browser platforms (1) do not allow libp2p implementations to
+manage the network sockets nor (2) do they allow managing the transport security
+stack, thus not allowing _insecure_ connections.
```

Comment:
I  find sending this doc useful when people want to understand what this "secure/insecure" division in browsers is about:

```suggestion
stack, thus not allowing _insecure_ connections from [secure contexts](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts).
```

Reply:
:+1: good idea. Included via https://github.com/libp2p/specs/pull/323/commits/7b494d6684a0edf3adae51080066ab15d3733797.